spi/src/main/java/org/opensearch/security/spi/resources/sharing/Recipients.java

```diff
@@ -60,16 +60,22 @@ public Map<Recipient, Set<String>> getRecipients() {
     public void share(Recipients target) {
         Map<Recipient, Set<String>> targetRecipients = target.getRecipients();
         for (Recipient recipientType : targetRecipients.keySet()) {
-            Set<String> updatedRecipients = recipients.get(recipientType);
-            updatedRecipients.addAll(targetRecipients.get(recipientType));
+            if (!recipients.containsKey(recipientType)) { // will not be present in case of very first share
```

Comment:
nit: This can be simplified to:

```
recipients
    .computeIfAbsent(recipientType, k -> new HashSet<>())
    .addAll(targetRecipients.getOrDefault(recipientType, Collections.emptySet()));
```